scripts/publish.ts

```diff
@@ -0,0 +1,408 @@
+      [...(await readCurrentBranchHeadTags())]
+        .map(parsePackageVersionSpecifier)
+        .filter(isDefined),
+    );
+    const prereleaseIdentifier = isOnDefaultBranch
+      ? maturityLevel
+      : `${maturityLevel}.${sanitizedCurrentBranchName}`;
+    packageNameToReleaseVersion = new Map(
+      await Promise.all(
+        packageNames.map(async (packageName) => {
+          const taggedVersion = packageNameToTaggedVersion.get(packageName);
+          if (taggedVersion !== undefined) {
+            // There's no need to increment the prerelease number if the HEAD
+            // already has a tagged version.
+            return [packageName, taggedVersion] as const;
+          }
+          const packageVersion = await readPackageVersion(packageName);
+          const parsedPackageVersion = semverParse(packageVersion);
+          if (parsedPackageVersion === null) {
+            throw new CommandError(
+              formatInvalidPackageVersionError(packageName, packageVersion),
+            );
+          }
+          const highestPublishedPrereleaseNumber =
+            findHighestPublishedPrereleaseNumber(
+              [...(packageNameToVersionHistory.get(packageName) ?? [])],
+              parsedPackageVersion,
+              prereleaseIdentifier,
+            );
+          const incrementedPackageVersion = incrementVersionByPrerelease(
+            parsedPackageVersion,
+            prereleaseIdentifier,
+            highestPublishedPrereleaseNumber,
+          );
+          return [packageName, incrementedPackageVersion] as const;
+        }),
+      ),
+    );
+  }
+  function formatInvalidPackageVersionError(
+    packageName: string,
+    packageVersion: string,
+  ): string {
+    return `Package \`${packageName}\` has an invalid version \`${packageVersion}\` in its manifest.`;
+  }
+
+  // Ensure the new versions aren't already in use.
+  if (!args["tolerate-republish"]) {
+    packageNameToReleaseVersion.forEach(
+      (packageReleaseVersion, packageName) => {
+        const packageVersionHistory =
+          packageNameToVersionHistory.get(packageName);
+        if (packageVersionHistory?.has(packageReleaseVersion)) {
+          throw new CommandError(
+            `Package \`${packageName}@${packageReleaseVersion}\` has already been published.`,
+          );
+        }
+      },
+    );
+  }
+
+  // Update the manifest of each package that is part of the release.
+  console.log(`✏️  Updating manifest${s}...`);
+  for (const nonPrivateWorkspaceName of nonPrivateWorkspaceNames) {
+    const nonPrivateWorkspaceReleaseVersion = packageNameToReleaseVersion.get(
+      nonPrivateWorkspaceName,
+    );
+    await exec(
+      "yarn",
+      [
+        ...["workspace", nonPrivateWorkspaceName],
+        ...["version", nonPrivateWorkspaceReleaseVersion ?? "decline"],
+        "--deferred",
+      ],
+      { dryRun },
+    );
+  }
+  await exec("yarn", ["version", "apply", "--all"], { dryRun });
```

Comment:
Is there a benefit to using `--deferred` rather than applying the versions immediately>